### PR TITLE
LogServer image bundling

### PR DESCRIPTION
**Pull request by kajes**
From appliance version 6.2, the LogServer function is decoupled from the base appliance image. Upgrading LogServer appliances will be done by downloading the required docker images for LogServer and uploading it to the Primary Controller, just like the base image is being uploaded already.

This PR will automatically pull the required LogServer images needed, pack them into a zip-file and upload it when running `upgrade prepare`.

Fixes: SA-21477

### CODE REVIEW

File: pkg/factory/http.go [hunk 1]
```diff
@@ -163,6 +166,51 @@ func customHTTPClient(f *Factory) func() (*http.Client, error) {
...
+			underlyingTransport: parentTransport,
+			username:            username,
+			password:            password,
+			useragent:           f.userAgent,
+			accept:              fmt.Sprintf("application/vnd.appgate.peer-v%d+json", f.Config.Version),
+		}
+		return client, nil
+	}
+}
+
```
Comment (danijeel) [inline pkg/factory/http.go:213]: Why is this needed? wouldn't it be enough to simply add the Authorization header key, value on the request?

File: .goreleaser.yml [hunk 2]
```diff
@@ -20,6 +20,9 @@ builds:
         - -s -w -X "github.com/appgate/sdpctl/cmd.version={{ if .Tag }}{{ .Tag }}{{ else }}dev{{ end }}"
         - -X "github.com/appgate/sdpctl/cmd.commit={{ .FullCommit }}"
         - -X "github.com/appgate/sdpctl/cmd.buildDate={{ .Date }}"
+        - -X "github.com/appgate/sdpctl/cmd/appliance/upgrade.dockerRegistry={{ .Env.DOCKER_REGISTRY_URL }}"
+        - -X "github.com/appgate/sdpctl/cmd/appliance/upgrade.dockerRegistryUsername={{ .Env.DOCKER_REGISTRY_USERNAME }}"
+        - -X "github.com/appgate/sdpctl/cmd/appliance/upgrade.dockerRegistryPassword={{ .Env.DOCKER_REGISTRY_PASSWORD }}"
```
Comment (danijeel) [inline .goreleaser.yml:25]: what is the reasoning behind this? why is these build vars? just curious...

File: pkg/appliance/functions.go [hunk 3]
```diff
@@ -776,3 +788,251 @@ func PrettyBytes(v float64) string {
...
+	manifestReq, err := http.NewRequestWithContext(ctx, http.MethodGet, manifestURL, nil)
+	if err != nil {
+		mainErrChan <- err
+		return
+	}
+	for k, v := range headers {
+		manifestReq.Header.Set(k, v)
+	}
+	manifestRes, err := client.Do(manifestReq)
+	if manifestRes.StatusCode != http.StatusOK || err != nil {
```
Comment (danijeel) [inline pkg/appliance/functions.go:869]: manifestRes can be null, should be enough to only check err
```suggestion
	if err != nil {
```

File: pkg/appliance/functions.go [hunk 4]
```diff
@@ -776,3 +788,251 @@ func PrettyBytes(v float64) string {
...
+	configReq, err := http.NewRequestWithContext(ctx, http.MethodGet, configURL, nil)
+	if err != nil {
+		mainErrChan <- err
+		return
+	}
+	for k, v := range headers {
+		configReq.Header.Set(k, v)
+	}
+	configRes, err := client.Do(configReq)
+	if configRes.StatusCode != http.StatusOK || err != nil {
```
Comment (danijeel) [inline pkg/appliance/functions.go:898]:
```suggestion
	if err != nil {
```
configRes can be null, should be enough to only check err

File: pkg/appliance/functions.go [hunk 5]
```diff
@@ -776,3 +788,251 @@ func PrettyBytes(v float64) string {
...
+		go func(l ManifestConfig, bufChan chan bufName, errChan chan error) {
+			f := strings.Replace(l.Digest, "sha256:", "", 1) + ".tar.gz"
+			layerURL := fmt.Sprintf("%s/v2/%s/blobs/%s", registry, image, l.Digest)
+			layerReq, err := http.NewRequestWithContext(ctx, http.MethodGet, layerURL, nil)
+			if err != nil {
+				errChan <- err
+				return
+			}
+			layerRes, err := client.Do(layerReq)
+			if layerRes.StatusCode != http.StatusOK || err != nil {
```
Comment (danijeel) [inline pkg/appliance/functions.go:929]:
```suggestion
			if err != nil {
```
layerRes can be null, avoid panic by only checking err

File: pkg/appliance/functions.go [hunk 6]
```diff
@@ -776,3 +788,251 @@ func PrettyBytes(v float64) string {
...
+		if err != nil {
+			errs = multierror.Append(errs, err)
+		}
+	}
+	log.Info("docker layers downloaded")
+
+	return archive, fileStat, errs.ErrorOrNil()
+}
+
+func downloadDockerImageBundle(ctx context.Context, out io.Writer, client *http.Client, ciMode bool, zipWriter *zip.Writer, registry, image, tag string, p *tui.Progress, mainErrChan chan<- error, wg *sync.WaitGroup) {
```
Comment (danijeel) [inline pkg/appliance/functions.go:851]: some inline code comments should be suitable for this func, with referance to docker docs.

File: cmd/appliance/upgrade/prepare.go [hunk 7]
```diff
@@ -34,14 +35,23 @@ import (
 	"github.com/hashicorp/go-version"
 	log "github.com/sirupsen/logrus"
 	"github.com/spf13/cobra"
-	"github.com/vbauerster/mpb/v7"
-	"github.com/vbauerster/mpb/v7/decor"
+)
+
+var (
+	// Set at build time or in env varibales
```
Comment (danijeel) [inline cmd/appliance/upgrade/prepare.go:41]:
```suggestion
	// Set at build time or in environment variables
```

File: pkg/appliance/functions.go [hunk 8]
```diff
@@ -776,3 +788,251 @@ func PrettyBytes(v float64) string {
...
+	Layers        []ManifestConfig
+}
+
+type ImageJSON struct {
+	Image string `json:"image"`
+}
+
+func DownloadDockerBundles(ctx context.Context, out io.Writer, client *http.Client, zipName, registry string, images map[string]string, ciMode bool) (*os.File, fs.FileInfo, error) {
+	// Create zip-archive
+	archive, err := os.Create(fmt.Sprintf("/tmp/%s.zip", zipName))
```
Comment (danijeel) [inline pkg/appliance/functions.go:810]: this wont work on windows, use https://pkg.go.dev/os#CreateTemp instead

File: pkg/appliance/functions.go [hunk 9]
```diff
@@ -776,3 +788,251 @@ func PrettyBytes(v float64) string {
...
+	}
+
+	log.Info("downloading image layers for ", zipName)
+	var p *tui.Progress
+	if !ciMode {
+		p = tui.New(ctx, out)
+		defer p.Wait()
+	}
+	errChan := make(chan error)
+	var wg sync.WaitGroup
```
Comment (danijeel) [inline pkg/appliance/functions.go:829]: https://pkg.go.dev/golang.org/x/sync/errgroup could work here as a option, that way we get rid of errChan, since we don't want to continue if any of these downloads gets  error. that would reduce complexity and func signature

File: pkg/factory/http.go (hunk 1 above)
Comment (kajes) [reply pkg/factory/http.go:213]: May be a bit overworked here, but it's mainly to make the check for authorization a bit cleaner in the prepare command, but now that I look at it, there's a bit of double checking the env variables too, so will fix that. Anyway, there's quite a lot of requests being made against the registry, so not having to deal with setting a header at each request was a conveniance, but I can simplify it if that's preferred.

File: pkg/appliance/functions.go (hunk 4 above)
Comment (kajes) [reply pkg/appliance/functions.go:898]: `401 Unauthorized` does not result in error though, which is a bit important if we want to support basic auth. Though the check should probably be reversed here if configRes can be null, so err should probably be checked first.

File: pkg/appliance/functions.go (hunk 3 above)
Comment (kajes) [reply pkg/appliance/functions.go:869]: same as configRes here

File: pkg/appliance/functions.go (hunk 8 above)
Comment (kajes) [reply pkg/appliance/functions.go:810]: :+1:

File: pkg/appliance/functions.go (hunk 4 above)
Comment (danijeel) [reply pkg/appliance/functions.go:898]: its still prone for panic, since configRes can be nil, you need to add a nested  if statement if you want to handle custom  statusCode check.

**Comment by thomascellerier** [comment]
I will do some manual testing on this and come back to you with some feedback.

**Comment by thomascellerier** [comment]
Now the bundling and uploading is working as expected.

We still need to validate that the contents of the bundle are in the format we expect. Which we can do after fixing some appliance side issues.

File: cmd/appliance/upgrade/prepare.go [hunk 10]
```diff
@@ -367,8 +408,98 @@ func prepareRun(cmd *cobra.Command, args []string, opts *prepareUpgradeOptions)
 		}
 	}
 
+	uploadWithProgress := func(ctx context.Context, reader io.Reader, name string, size int64, headers map[string]string) error {
+		p := tui.New(ctx, spinnerOut)
+		defer p.Wait()
+		endMsg := "uploaded"
+		proxyReader, t := p.FileUploadProgress(name, endMsg, size, reader)
```
Comment (danijeel) [inline cmd/appliance/upgrade/prepare.go:415]: looks like you are missing `proxyReader.Close()`

File: cmd/appliance/upgrade/prepare.go [hunk 11]
```diff
@@ -731,8 +819,9 @@ func prepareRun(cmd *cobra.Command, args []string, opts *prepareUpgradeOptions)
 
 const prepareUpgradeMessage = `PREPARE SUMMARY
 
-1. Upload upgrade image {{.Filepath}} to Controller
-2. Prepare upgrade on the following appliances:
+1. Bundle and upload LogServer docker image if needed
```
Comment (danijeel) [inline cmd/appliance/upgrade/prepare.go:822]: Should this always be shown, even though there are on log server in the collective?

Comment (thomascellerier) [reply cmd/appliance/upgrade/prepare.go:822]: No this should only be shown if there is a log server being upgraded.
We should make sure the whole flow is only triggered in that case.

File: cmd/appliance/upgrade/prepare.go (hunk 10 above)
Comment (kajes) [reply cmd/appliance/upgrade/prepare.go:415]: Yep, I'll fix that :+1:

File: cmd/appliance/upgrade/prepare.go (hunk 11 above)
Comment (kajes) [reply cmd/appliance/upgrade/prepare.go:822]: Probably not, but takes a bit of work to avoid it, so I guess the question would be if it's a deal breaker before merging?

Comment (thomascellerier) [reply cmd/appliance/upgrade/prepare.go:822]: I think so yes, this will confuse people.
It also needs to be only shown for upgrades to 6.2.0+ (with a logserver being upgraded).

Comment (kajes) [reply cmd/appliance/upgrade/prepare.go:822]: Ok, I'll give it some time then before merge

File: cmd/appliance/upgrade/prepare.go [hunk 12]
```diff
@@ -367,8 +408,98 @@ func prepareRun(cmd *cobra.Command, args []string, opts *prepareUpgradeOptions)
...
+				err = uploadWithProgress(ctx, pr, "uploading "+logServerZipName, zipInfo.Size(), headers)
+			} else {
+				err = a.UploadFile(ctx, pr, headers)
+				fmt.Fprint(opts.Out, "Image bundles prepared\n\n")
+			}
+			if err != nil {
+				return err
+			}
+			os.Remove(zipPath)
+			fmt.Fprintln(opts.Out)
```
Comment (danijeel) [inline cmd/appliance/upgrade/prepare.go:496]:
```suggestion
```

**Review by thomascellerier — APPROVED** [review]
